lib/runtime/wazero/imports.go

```diff
@@ -0,0 +1,2247 @@
+	if signingKey == nil {
+		logger.Error("could not find public key " + pubKey.Hex() + " in keystore")
+		ret, err := write(m, rtCtx.Allocator, []byte{0})
+		if err != nil {
+			panic(err)
+		}
+		return ret
+	}
+
+	sig, err := signingKey.Sign(read(m, msg))
+	if err != nil {
+		logger.Error("could not sign message")
+	}
+
+	var fixedSize [64]byte
+	copy(fixedSize[:], sig)
+
+	encoded, err := scale.Marshal(&fixedSize)
+	if err != nil {
+		logger.Error(fmt.Sprintf("scale encoding: %s", err))
+		ret, err := write(m, rtCtx.Allocator, []byte{0})
+		if err != nil {
+			panic(err)
+		}
+		return ret
+	}
+
+	ret, err := write(m, rtCtx.Allocator, encoded)
+	if err != nil {
+		logger.Errorf("failed to allocate memory: %s", err)
+		return 0
+	}
+
+	return ret
+}
+
+func ext_crypto_ed25519_verify_version_1(ctx context.Context, m api.Module, sig uint32, msg uint64, key uint32) uint32 {
+	rtCtx := ctx.Value(runtimeContextKey).(*runtime.Context)
+	if rtCtx == nil {
+		panic("nil runtime context")
+	}
+
+	memory := m.Memory()
+	sigVerifier := rtCtx.SigVerifier
+
+	// signature := memory[sig : sig+64]
+	signature, ok := memory.Read(sig, 64)
+	if !ok {
+		panic("read overflow")
+	}
+	message := read(m, msg)
+	pubKeyData, ok := memory.Read(key, 32)
+	if !ok {
+		panic("read overflow")
+	}
+
+	pubKey, err := ed25519.NewPublicKey(pubKeyData)
+	if err != nil {
+		logger.Error("failed to create public key")
+		return 0
+	}
+
+	if sigVerifier.IsStarted() {
+		signature := crypto.SignatureInfo{
+			PubKey:     pubKey.Encode(),
+			Sign:       signature,
+			Msg:        message,
+			VerifyFunc: ed25519.VerifySignature,
+		}
+		sigVerifier.Add(&signature)
+		return 1
+	}
+
+	if ok, err := pubKey.Verify(message, signature); err != nil || !ok {
+		logger.Error("failed to verify")
+		return 0
+	}
+
+	logger.Debug("verified ed25519 signature")
+	return 1
+}
+
+func ext_crypto_secp256k1_ecdsa_recover_version_1(ctx context.Context, m api.Module, sig, msg uint32) uint64 {
+	rtCtx := ctx.Value(runtimeContextKey).(*runtime.Context)
+	if rtCtx == nil {
+		panic("nil runtime context")
+	}
+
+	// msg must be the 32-byte hash of the message to be signed.
+	// sig must be a 65-byte compact ECDSA signature containing the
+	// recovery id as the last element
+	message, ok := m.Memory().Read(msg, 32)
+	if !ok {
+		panic("read overflow")
+	}
+	signature, ok := m.Memory().Read(sig, 65)
+	if !ok {
+		panic("read overflow")
+	}
+
+	res := scale.NewResult([64]byte{}, nil)
+
+	pub, err := secp256k1.RecoverPublicKey(message, signature)
+	if err != nil {
+		logger.Errorf("failed to recover public key: %s", err)
+		err := res.Set(scale.Err, nil)
+		if err != nil {
+			panic(err)
+		}
+		ret, err := write(m, rtCtx.Allocator, scale.MustMarshal(res))
+		if err != nil {
+			panic(err)
+		}
+		return ret
+	}
+
+	logger.Debugf(
+		"recovered public key of length %d: 0x%x",
+		len(pub), pub)
+
+	var fixedSize [64]byte
+	copy(fixedSize[:], pub[1:])
+
+	err = res.Set(scale.OK, fixedSize)
```

Comment:
Why is this fixed sized/do we need to account for case where data is empty (err result case)?

Reply:
I'm using `scale.Result` and on line 345, the  error case is empty.  I'm using a fixed size here, since that's what's expected from the runtime.  AKA, it shouldn't be length encoded.